packages/mask/src/plugins/Referral/SNSAdaptor/MyFarms.tsx

```diff
@@ -0,0 +1,336 @@
+import { Grid, Typography, CircularProgress, Button, Box } from '@mui/material'
+import { EMPTY_LIST } from '@masknet/shared-base'
+
+import { useI18N } from '../../../utils'
+import { toNativeRewardTokenDefn, parseChainAddress, roundValue, makeLeafHash, getRequiredChainId } from '../helpers'
+import {
+    Farm,
+    EntitlementLog,
+    PageInterface,
+    PagesType,
+    TransactionStatus,
+    RewardsHarvestedEvent,
+    TabsReferralFarms,
+} from '../types'
+import { ReferralRPC } from '../messages'
+import { harvestRewards } from './utils/referralFarm'
+
+import { AccordionFarm } from './shared-ui/AccordionFarm'
+import { EthereumChainBoundary } from '../../../web3/UI/EthereumChainBoundary'
+
+import { useSharedStyles } from './styles'
+
+const useStyles = makeStyles()((theme) => ({
+    container: {
+        lineHeight: '22px',
+        fontWeight: 300,
+        '& > div::-webkit-scrollbar': {
+            width: '7px',
+        },
+        '& > div::-webkit-scrollbar-track': {
+            boxShadow: 'inset 0 0 6px rgba(0,0,0,0.00)',
+            webkitBoxShadow: 'inset 0 0 6px rgba(0,0,0,0.00)',
+        },
+        '& > div::-webkit-scrollbar-thumb': {
+            borderRadius: '4px',
+            backgroundColor: theme.palette.background.default,
+        },
+    },
+    col: {
+        color: theme.palette.text.secondary,
+        fontWeight: 500,
+    },
+    heading: {
+        paddingRight: '27px',
+    },
+    content: {
+        height: 320,
+        overflowY: 'scroll',
+        marginTop: 20,
+        color: theme.palette.text.strong,
+        display: 'flex',
+        flexDirection: 'column',
+        alignItems: 'center',
+    },
+    accordion: {
+        width: '100%',
+    },
+    accordionSummary: {
+        margin: 0,
+        padding: 0,
+    },
+    noFarm: {
+        width: '100%',
+        display: 'flex',
+        alignItems: 'center',
+        justifyContent: 'center',
+        borderRadius: '12px',
+        background: theme.palette.background.default,
+        height: '44px',
+        color: theme.palette.text.strong,
+        fontWeight: 500,
+    },
+    total: {
+        marginRight: '5px',
+    },
+    button: {
+        marginLeft: 'auto',
+    },
+}))
+
+interface FarmsListProps extends PageInterface {
+    entitlements: EntitlementLog[]
+    rewardsHarvested: RewardsHarvestedEvent[]
+    allTokens: ERC20TokenDetailed[]
+    farms: Farm[]
+}
+function FarmsList({ entitlements, allTokens, farms, rewardsHarvested, ...props }: FarmsListProps) {
+    const { t } = useI18N()
+    const currentChainId = useChainId()
+    const account = useAccount()
+    const web3 = useWeb3({ chainId: currentChainId })
+    const { value: nativeToken } = useNativeTokenDetailed()
+    const { showSnackbar } = useCustomSnackbar()
+    const pageType = props?.pageType || PagesType.REFERRAL_FARMS
+
+    const allTokensMap = new Map(allTokens.map((token) => [token.address.toLowerCase(), token]))
```

Comment:
Attention: These complex objects are updated every time this component got rendered.

Reply:
Good point, refactoring done.

Reply:
Good point, refactoring done.